### PR TITLE
Penalty updates

### PR DESCRIPTION
**Pull request by p-ferreira**
- adds `MatchContentCriteria`
- adds `SimpleResponseLayoutCriteria`
- adds `MatchLayoutCriteria`
- modifies following pruning at forward step
- complement patterns in sentence match penalty function
- adapt current flow to incorporate new criteria
- adapt max penalties

### CODE REVIEW

File: prompting/validators/criteria.py [hunk 1]
```diff
@@ -109,3 +109,103 @@ def evaluate(self, completions: List[str]) -> torch.FloatTensor:
...
+        # Adjust the text based on the number of words
+        if self.n_words > 1:
+            text = f"Your response {should_match_text} {match_type_text} one of the following words: {words_list}."
+        else:
+            text = f"Your response {should_match_text} {match_type_text} the following word: {words_list}."
+        return text
+
+
+@dataclass
+class SimpleResponseLayoutCriteria(TaskCriterion):
```
Comment (steffencruz) [inline prompting/validators/criteria.py:190]: WIP?

File: prompting/validators/tasks.py [hunk 2]
```diff
@@ -155,35 +202,46 @@ def create_qg_task(base_text: str, index: int) -> QuestionGenerationTask:
...
-            target_length=random.randint(50, 200),
-            unit=TextLengthUnitEnum.WORDS,
-        ),
-        MatchLengthCriteria(
-            penalty=0.1,
-            target_length=random.randint(4, 8),
-            unit=TextLengthUnitEnum.SENTENCES,
-        ),
-    ]
+    answer_should_not_include_criteria = MatchContentCriteria(
```
Comment (steffencruz) [inline prompting/validators/tasks.py:220]: what does n_words mean in this context? The probability of this criteria being applied?

**Review by steffencruz — COMMENTED** [review]
MatchContentCriteria looks great. Looking forward to seeing MatchLayoutCriteria

File: prompting/validators/tasks.py (hunk 2 above)
Comment (p-ferreira) [reply prompting/validators/tasks.py:220]: The amount of words you want to sample, in order to not bloat the prompt with an extensive list of possibilities. So for the question generation for example, we provide a list of possible `question_prefixes` to the criteria and the `n_words` will random select from this list n samples to be considered in the prompt.

We could drop this parameter and do the sample outside of the criteria as well

Comment (steffencruz) [reply prompting/validators/tasks.py:220]: no i like it. so with a value of 0.2 it will select 20% of the elements of the list?

**Review by steffencruz — APPROVED** [review]
LGTM

**Pull request merged**